src/lib/arch/riscv/Instruction_decode.cc

```diff
@@ -0,0 +1,254 @@
+    exceptionEncountered_ = true;
+    return;
+  }
+
+  // Identify branches
+  switch (metadata.opcode) {
+    case Opcode::RISCV_BEQ:
+    case Opcode::RISCV_BNE:
+    case Opcode::RISCV_BLT:
+    case Opcode::RISCV_BLTU:
+    case Opcode::RISCV_BGE:
+    case Opcode::RISCV_BGEU:
+    case Opcode::RISCV_JAL:
+    case Opcode::RISCV_JALR:
+      isBranch_ = true;
+      break;
+      // Identify loads/stores
+    case Opcode::RISCV_LR_D:
+    case Opcode::RISCV_LR_D_AQ:
+    case Opcode::RISCV_LR_D_RL:
+    case Opcode::RISCV_LR_D_AQ_RL:
+    case Opcode::RISCV_LR_W:
+    case Opcode::RISCV_LR_W_AQ:
+    case Opcode::RISCV_LR_W_RL:
+    case Opcode::RISCV_LR_W_AQ_RL:
+      isAtomic_ = true;
+      [[fallthrough]];
+    case Opcode::RISCV_LB:
+    case Opcode::RISCV_LBU:
+    case Opcode::RISCV_LH:
+    case Opcode::RISCV_LHU:
+    case Opcode::RISCV_LW:
+    case Opcode::RISCV_LWU:
+    case Opcode::RISCV_LD:
+      isLoad_ = true;
+      break;
+    case Opcode::RISCV_SC_D:
+    case Opcode::RISCV_SC_D_AQ:
+    case Opcode::RISCV_SC_D_RL:
+    case Opcode::RISCV_SC_D_AQ_RL:
+    case Opcode::RISCV_SC_W:
+    case Opcode::RISCV_SC_W_AQ:
+    case Opcode::RISCV_SC_W_RL:
+    case Opcode::RISCV_SC_W_AQ_RL:
+      isAtomic_ = true;
+      [[fallthrough]];
+    case Opcode::RISCV_SB:
+    case Opcode::RISCV_SW:
+    case Opcode::RISCV_SH:
+    case Opcode::RISCV_SD:
+      isStore_ = true;
+      break;
+  }
+
+  if (Opcode::RISCV_AMOADD_D <= metadata.opcode &&
+      metadata.opcode <= Opcode::RISCV_AMOXOR_W_RL) {
+    // Atomics: both load and store
+    isLoad_ = true;
+    isStore_ = true;
+    isAtomic_ = true;
+  }
+
+  // Extract explicit register accesses, ignore immediates until execute
+  for (size_t i = 0; i < metadata.operandCount; i++) {
+    const auto& op = metadata.operands[i];
+
+    // First operand is always be of REG type but could be either source or
+    // destination
+    if (i == 0 && op.type == RISCV_OP_REG) {
+      // If opcode is branch or store (but not atomic or jump) the first operand
+      // is a source register, for all other instructions the first operand is a
+      // destination register
+      if ((isBranch() && metadata.opcode != Opcode::RISCV_JAL &&
+           metadata.opcode != Opcode::RISCV_JALR) ||
+          (isStoreAddress() && !isAtomic())) {
+        sourceRegisters[sourceRegisterCount] = csRegToRegister(op.reg);
+
+        if (sourceRegisters[sourceRegisterCount] ==
+            Instruction::ZERO_REGISTER) {
+          // Catch zero register references and pre-complete those operands
+          operands[sourceRegisterCount] = RegisterValue(0, 8);
+        } else {
+          operandsPending++;
+        }
+
+        sourceRegisterCount++;
+      } else {
+        if (csRegToRegister(op.reg) != Instruction::ZERO_REGISTER) {
+          destinationRegisters[destinationRegisterCount] =
+              csRegToRegister(op.reg);
+
+          destinationRegisterCount++;
+        }
```

Comment:
Why is there no defined behaviour for Zero-registers?

Reply:
Comment written